### PR TITLE
[BREAKING] deprecate name => fun in favor of fun => name in describe

### PR DESCRIPTION
**Pull request by bkamins**
Fixes https://github.com/JuliaData/DataFrames.jl/issues/2385

### CODE REVIEW

File: src/abstractdataframe/abstractdataframe.jl [hunk 1]
```diff
@@ -530,19 +530,25 @@ julia> describe(df, :min, :sum => sum)
...
-DataAPI.describe(df::AbstractDataFrame,
-                 stats::Union{Symbol, Pair{<:SymbolOrString}}...;
-                 cols=:) =
-    _describe(select(df, cols, copycols=false), collect(stats))
-
+function DataAPI.describe(df::AbstractDataFrame,
+                          stats::Union{Symbol,
+                                       Pair{<:Base.Callable,<:SymbolOrString},
+                                       Pair{<:SymbolOrString}}...; # TODO: remove after deprecation
+                 cols=:)
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:545]: Align with previous line.

File: src/abstractdataframe/abstractdataframe.jl [hunk 2]
```diff
@@ -530,19 +530,25 @@ julia> describe(df, :min, :sum => sum)
...
+    if any(x -> x isa Pair{<:SymbolOrString}, stats)
+        Base.depwarn("name => function order is deprecated; use function => name instead", :describe)
+    end
+    return _describe(select(df, cols, copycols=false),
+           [s isa Pair{<:SymbolOrString} ? last(s) => first(s) : s for s in stats])
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:550]: Align too.

**Review by nalimilan — APPROVED** [review]
Looks good with fixed indentation!

File: src/abstractdataframe/abstractdataframe.jl (hunk 1 above)
Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl:545]: done

File: src/abstractdataframe/abstractdataframe.jl (hunk 2 above)
Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl:550]: done

**Comment by bkamins** [comment]
@matthieugomez - what I changed here is what we all agree to change.

Now - do you have strong preference for your proposal? If yes, just to make sure, you envision the signature to be:
```
describe(df, cols=:; stats=[:mean, :min, :median, :max, :nmissing, :eltype])
```
where stats is a single stat or a vector of statistics to calculate?

**Comment by matthieugomez** [comment]
I would prefer it but it's not the hill I'm ready to die on ;) It's
interesting to hear you were using it in a different way.

**Comment by bkamins** [comment]
@nalimilan - do you have any preference here?

**Comment by nalimilan** [comment]
I don't have a strong opinion. At least what is clear is that functions which take a `cols` (positional) argument only allow a single one in general, but functions that take `source => fun => dest` allow varargs. So in that sense the current situation is consistent.

**Comment by bkamins** [comment]
Thank you for considering it with such care. Given all pros and cons I am leaning to leave the PR as it stands now.

**Comment by bkamins** [comment]
@nalimilan - so do we go for this or wait a bit for more feedback (I feel all has been said already).

File: NEWS.md [hunk 3]
```diff
@@ -32,6 +32,12 @@
   choose the fast path only when it is safe; this resolves inconsistencies
   with what the same functions not using fast path produce
   ([#2357](https://github.com/JuliaData/DataFrames.jl/pull/2357))
+* the `categorical` and `categorical!` functions have been deprecated in favor of
+  `transform(df, cols .=> categorical .=> cols)` and similar syntaxes
+  ([#2394](https://github.com/JuliaData/DataFrames.jl/pull/2394))
```
Comment (nalimilan) [inline NEWS.md:37]:
```suggestion
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 4]
```diff
@@ -530,19 +530,24 @@ julia> describe(df, :min, :sum => sum)
 │ 2   │ x        │ 0.1 │ 5.5 │
 │ 3   │ y        │ 'a' │     │
 
-julia> describe(df, :min, :sum => sum, cols=:x)
+julia> describe(df, cols=:x, :min, sum => :sum)
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:533]: I'd rather keep `cols` at the end, that's more common.
```suggestion
julia> describe(df, :min, sum => :sum, cols=:x)
```

File: src/abstractdataframe/abstractdataframe.jl [hunk 5]
```diff
@@ -530,19 +530,24 @@ julia> describe(df, :min, :sum => sum)
...
-
+function DataAPI.describe(df::AbstractDataFrame, stats::Union{Symbol,
+                          Pair{<:Base.Callable,<:SymbolOrString},
+                          Pair{<:SymbolOrString}}...; # TODO: remove after deprecation
+                 cols=:)
+    if any(x -> x isa Pair{<:SymbolOrString}, stats)
+        Base.depwarn("name => function order is deprecated; use function => name instead", :describe)
+    end
+    return _describe(select(df, cols, copycols=false),
+                     [s isa Pair{<:SymbolOrString} ? last(s) => first(s) : s for s in stats])
```
Comment (nalimilan) [inline src/abstractdataframe/abstractdataframe.jl:549]: Maybe better for latency?
```suggestion
                     Any[s isa Pair{<:SymbolOrString} ? last(s) => first(s) : s for s in stats])
```

File: test/dataframe.jl [hunk 6]
```diff
@@ -683,6 +683,9 @@ end
     @test describe(df, cols=Not(1)) ≅ describe(select(df, Not(1)))
     @test describe(df, cols=Not("a")) ≅ describe(select(df, Not(1)))
 
+    @test describe(DataFrame(a=[1,2]), cols = :a, :min, minimum => :min2, maximum => "max2", :max) ==
+          DataFrame(variable=:a, min=1, min2=1, max2=2, max=2)
+
     @test_throws ArgumentError describe(df, :mean, :all)
 end
```
Comment (nalimilan) [inline test/dataframe.jl:690]: The deprecation also needs a test.

**Comment by nalimilan** [comment]
Yeah, let's go with this since there's no perfect solution.

File: src/abstractdataframe/abstractdataframe.jl (hunk 4 above)
Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl:533]: I made it on purpose as @matthieugomez wanted the `cols` argument to go second (and kwarg can be placed anywhere). So - which style do we want to settle to?

File: src/abstractdataframe/abstractdataframe.jl (hunk 5 above)
Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl:549]: and safer in general :smile:

File: src/abstractdataframe/abstractdataframe.jl (hunk 4 above)
Comment (bkamins) [reply src/abstractdataframe/abstractdataframe.jl:533]: (I am accepting the suggestion, but we might revert it if we want)

Comment (nalimilan) [reply src/abstractdataframe/abstractdataframe.jl]: I mean it's fine if people want to put it first (that's probably a good idea), but for clarity in the reference docs I'd put it last, like we always do for keyword arguments.

File: test/dataframe.jl (hunk 6 above)
Comment (bkamins) [reply test/dataframe.jl:690]: added

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**